### DISCUSSION TITLE
[Enhancement] Build tools used

### DISCUSSION THREAD

**Discussion by LetMeR00t** [mitchos/pyZscaler#127] [General]
Hi,
This issue is not related to a bug/feature about pyZscaler but more to add information about how you are currently developing this tool.
In fact, I never created a python library on my own but I want to work on a python library not yet available and I want to inspire from your way of developing because I found your library very clear and well developed.

Which tools are you using to develop this library ? I found some names like "Poetry" but it's difficult for me to understand what is the usage of each tool ... So if you have time to share your way of developing and tools, it would be really appreciated.

Thank you in advance

**Comment by mitchos** [comment]
Hey,

Thanks for reaching out. Appreciate your kind words. I was driven to develop this library in this particular way due to the huge amount of poorly documented libraries and SDKs. 

### IDE:
- [Pycharm](https://www.jetbrains.com/pycharm/) with the following plugins
  - Gitflow Integration Plus
  - Material Theme UI

### Dev Libraries
- [restfly](https://github.com/librestfly/restfly): A great simple framework for building libraries that interact with REST APIs
- [sphinx](https://www.sphinx-doc.org/en/master/): Documentation
- [furo](https://pypi.org/project/furo/): I think it's a nice doc theme for Sphinx
- [python box:](https://pypi.org/project/python-box/) Found this when working with Restfly, not entirely necessary but it's nice to have dot notation access to object properties
- [responses](https://pypi.org/project/responses/): Used for mocking tests in Pytest
- [pytest](https://docs.pytest.org/en/7.1.x/): Used for the test suite
- [pre-commit:](https://pre-commit.com/) great library for ensuring certain actions take place before commit
- [black](https://github.com/psf/black): I like this opinionated formatter, I use this in all my projects now
- [flake8](https://flake8.pycqa.org/en/latest/): pretty standard for python linting
- [isort](https://pycqa.github.io/isort/): sorts imports across all your modules
- [poetry](https://python-poetry.org/): packaging and dependency management

### Testing
I write tests as I go, ensuring that all conditional branches are covered. As the library has gotten bigger, it's crucial to ensuring things like dependency updates don't break.

The only problem is I don't have a test suite to ensure that Zscaler's API hasn't changed. I believe there's been one or two occasions where an API endpoint has changed without being re-versioned, which can be frustrating.

### Pre-commit 
In pre-commit I use 

- flake8 to lint my code. I really prefer a wider 127 line format. It's 2022, I have an ultra-wide screen and so should you :)
- Black is used for formatting
- isort for imports
- Still looking to see if I should add anything additional to the pre-commit workflow. 

### Git
- I like to use the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard for my commit messages. Sure I slip every now and then but for the most part it keeps everything pretty tidy. 
- I like to use the Github syntax of 'closes #100' or 'resolves #101' or 'fixes #102' to close issues/PRs when merging a release branch into main.
- I use [Gitflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) as I find it well structured and easy to follow. The Gitflow plugin for Pycharm makes life easy for creating and merging feature, hotfix and bugfix branches.

### Release flow
1. I merge all my feature branches in develop
2. Create release branch from develop
3. Tidy anything up and update the version (need to automate this using a version bump tool, doing it manually at the moment)
4. Merge release into develop and main
5. Push develop and main to Github
6. Poetry build package
7. Poetry publish to PyPI
8. Create release notes on Github

I hope that helps. There's probably a million better ways to do this but this is working great for me.

**Comment by LetMeR00t** [comment]
Hi
Thank you so much for your feedback.
It’s really interesting and educative.
As you mentioned there are a lot of ways to do such things but at least the output of it is important and when I see how the library is developed, I’m in tune with your way of working.
I used to work with Visual Studio because it’s more generic in fact.
I never heard about conventional commits and Gitflow which seems very interesting to know and follow, thank you for that.
For the rest, I already started to understand how it works and it seems to be easy to use.
Again thank you for your feedback, I will follow your way of working to develop the new library (regarding Checkpoint firewalls products)
Have a good day

**Comment by LetMeR00t** [comment]
Hi @mitchos ,
I tried to configure a workflow with pytest in Github (action) but I'm stuck with an error I can't understand ..

You'll find an example here: https://github.com/LetMeR00t/pyCheckpoint/runs/7212486556?check_suite_focus=true

![image](https://user-images.githubusercontent.com/10530375/177531539-6b9989c1-0192-48b6-8deb-c2a4a7b32ae8.png)

It seems that my own module isn't detected by pytest. On my workstation, I'm using poetry so the package is built locally within a dedicated environment (and the tests are running well) but it's not the same with Github Action.
I've check with your own code and I didn't notice what would be the cause as we seems to be doing the same things.

Did you already have this issue before by any chance ?

Thank you

**Comment by LetMeR00t** [comment]
Hi @mitchos ,
FYI, I followed this post here: https://github.com/pytest-dev/pytest/issues/2287
Adding "pip install -e ." solved my issue.
